packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/CachedQueryService.kt

```diff
@@ -0,0 +1,83 @@
+package com.mongodb.jbplugin.editor
+
+import com.intellij.database.dataSource.LocalDataSource
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiManager
+import com.intellij.psi.util.CachedValue
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiTreeUtil
+import com.mongodb.jbplugin.accessadapter.datagrip.DataGripBasedReadModelProvider
+import com.mongodb.jbplugin.accessadapter.datagrip.adapter.isConnected
+import com.mongodb.jbplugin.accessadapter.slice.BuildInfo
+import com.mongodb.jbplugin.mql.Node
+import com.mongodb.jbplugin.mql.components.HasTargetCluster
+import io.github.z4kn4fein.semver.Version
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * Attempts to parse a query for a PsiElement if there is one, given the dialect of the current
+ * file. Caches the result inside the PsiElement metadata and only reparses the query if that
+ * element or it's children change.
+ *
+ * It might return null if there is no query to parse.
+ *
+ * @property coroutineScope
+ */
+@Service(Service.Level.PROJECT)
+class CachedQueryService(
```

Comment:
We had duplicated code across the bridges, and each bridge was parsing the query once and then caching it, that is a waste of resources. Now the cached query service is able to reuse queries parsed across different bridges and reuse them.